### PR TITLE
Refactor context so its has less hacks

### PR DESCRIPTION
**Pull request by whatknight**
addresses one of the things for 4.0.0 release https://github.com/jdlehman/switcheroo/issues/29

### CODE REVIEW

File: src/SwitcherProvider.js [hunk 1]
```diff
@@ -26,15 +23,33 @@ export default function SwitcherProvider(props) {
     };
   }, []);
 
+  const providedMethods = useMemo(() => {
```
Comment (jdlehman) [inline src/SwitcherProvider.js]: why not `useCallback` here?

File: src/SwitcherProvider.js [hunk 2]
```diff
@@ -26,15 +23,33 @@ export default function SwitcherProvider(props) {
...
+    const getListenerList = type => {
+      switch (type) {
+        case 'load':
+          return loadListeners.current;
+        case 'popstate':
+          return popStateListeners.current;
+        case 'hashchange':
+          return hashChangeListeners.current;
+        default:
+          throw new Error('womp womp');
```
Comment (jdlehman) [inline src/SwitcherProvider.js]: nit: lets make this error more helpful even if it technically should never be hit.

File: src/helpers.js [hunk 3]
```diff
@@ -1,14 +1,5 @@
 import { Children } from 'react';
 
-// http://stackoverflow.com/a/2117523
-export function generateGuid() {
```
Comment (jdlehman) [inline src/helpers.js]: YES!

Comment (jdlehman) [reply src/helpers.js]: I won't miss any of this code

File: src/Switcher.js [hunk 4]
```diff
@@ -81,47 +81,26 @@ function Switcher(props) {
   useEffect(() => {
     const handler = () => handleRouteChange.current();
     const usingProvider = Boolean(switcherProvider);
-    if (!usingProvider) {
-      load && window.addEventListener('load', handler);
-      pushState && window.addEventListener('popstate', handler);
-      hashChange && window.addEventListener('hashchange', handler);
+    if (usingProvider) {
```
Comment (jdlehman) [inline src/Switcher.js]: this inversion makes the code so much better! I love it.

What do you think of ALWAYS requiring the provider? such that the second case is never even possible?

I think lots of other libraries do this. Might be worth making some of these breaking changes before the next release.

File: src/SwitcherProvider.js (hunk 1 above)
Comment (whatknight) [reply src/SwitcherProvider.js]: that was my initial implementation actually. but it seemed better to memo the object passed into the provider so it wouldn't cause the consumers to rerender if the provider rerendered. so i just stuck everything in one memo to make it simpler.

File: src/SwitcherProvider.js (hunk 2 above)
Comment (whatknight) [reply src/SwitcherProvider.js]: lol that's not a nit. thats just a good pr comment

File: src/Switcher.js (hunk 4 above)
Comment (whatknight) [reply src/Switcher.js]: it is kind of nice to be able to throw in one component but i think i agree it doesn't scale well if i have multiple switchers rendered in one tree. but i'm not sure how often that case would occur.

Comment (jdlehman) [reply src/Switcher.js]: cool, we won't do that in this PR, but maybe before the release

**Pull request merged**